File: test/cpu/cas.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_CAS {
    F_CAS() {
        // CAS.B %D4, %D3, (%A2)
        TEST::SET_W(0, 0005320 | 2);
        TEST::SET_W(2, 3 << 6 | 4);
        TEST::SET_W(4, TEST_BREAK);

        // CAS.W %D4, %D3, (%A2)
        TEST::SET_W(6, 0006320 | 2);
        TEST::SET_W(8, 3 << 6 | 4);
        TEST::SET_W(10, TEST_BREAK);

        // CAS.L %D4, %D3, (%A2)
        TEST::SET_W(12, 0007320 | 2);
        TEST::SET_W(14, 3 << 6 | 4);
        TEST::SET_W(16, TEST_BREAK);

        jit_compile(0, 18);
    }};
BOOST_FIXTURE_TEST_SUITE(CAS, Prepare, *boost::unit_test::fixture<F_CAS>())


BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(T) {
    RAM[0x100] = 0x22;
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x30;
    cpu.D[4] = 0x22;
    run_test(0);
    BOOST_TEST(RAM[0x100] == 0x30);
}

BOOST_AUTO_TEST_CASE(F) {
    RAM[0x100] = 0x22;
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x30;
    cpu.D[4] = 0x35;
    run_test(0);
    BOOST_TEST(cpu.D[4] == 0x22);
}

BOOST_AUTO_TEST_CASE(traced) {
    cpu.T = 1;
	run_test(0);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::TRACE );
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(T) {
    TEST::SET_W(0x100, 0x2222);
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x3030;
    cpu.D[4] = 0x2222;
    run_test(6);
    BOOST_TEST(TEST::GET_W(0x100) == 0x3030);
}

BOOST_AUTO_TEST_CASE(F) {
    TEST::SET_W(0x100, 0x2222);
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x3030;
    cpu.D[4] = 0x3535;
    run_test(6);
    BOOST_TEST(cpu.D[4] == 0x2222);
}

BOOST_AUTO_TEST_CASE(traced) {
    cpu.T = 1;
	run_test(6);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::TRACE );
}

BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(T) {
    TEST::SET_L(0x100, 0x22222222);
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x30303030;
    cpu.D[4] = 0x35353535;
    run_test(12);
    BOOST_TEST(cpu.D[4] == 0x22222222);
}

BOOST_AUTO_TEST_CASE(F) {
    TEST::SET_L(0x100, 0x22222222);
    cpu.A[2] = 0x100;
    cpu.D[3] = 0x30303030;
    cpu.D[4] = 0x22222222;
    run_test(12);
    BOOST_TEST(TEST::GET_L(0x100) == 0x30303030);
}

BOOST_AUTO_TEST_CASE(traced) {
    cpu.T = 1;
	run_test(12);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::TRACE );
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/cas.lua
-- CAS.B %D1, %D2, (%A3)
MEM[0].W = OCT"0005320" | 3
MEM[2].W = 2 << 6 | 1
MEM[4].W = TEST.BREAK

-- CAS.W %D2, %D3, (%A4)
MEM[6].W = OCT"0006320" | 4
MEM[8].W = 3 << 6 | 2
MEM[10].W = TEST.BREAK

-- CAS.L %D3, %D4, (%A5)
MEM[12].W = OCT"0007320" | 5
MEM[14].W = 4 << 6 | 3
MEM[16].W = TEST.BREAK

JIT_COMPILE(0, 18)

local run = {}

run.B = {
   pre = function()
      MEM[0x1000].B = 0x22
      CPU.A[4] = 0x1000
      CPU.D[3].B = 0x30
   end,
   t = function()
      CPU.D[2].B = 0x22
      TEST.RUN(0)
      TEST.CHECK("MEM[0x1000].B", 0x30)
   end,
   f = function()
      CPU.D[2].B = 0x55
      TEST.RUN(0)
      TEST.CHECK("CPU.D[2].B", 0x22)
   end,
   trace = function()
      CPU.A[4] = 0x1000
      trace_test(0)
   end,
}

run.W = {
   pre = function()
      MEM[0x1000].W = 0x2222
      CPU.A[5] = 0x1000
      CPU.D[4].W = 0x3030
   end,
   t = function()
      CPU.D[3].W = 0x2222
      TEST.RUN(6)
      TEST.CHECK("MEM[0x1000].W", 0x3030)
   end,
   f = function()
      CPU.D[3].W = 0x5555
      TEST.RUN(6)
      TEST.CHECK("CPU.D[3].W", 0x2222)
   end,
   trace = function()
      CPU.A[5] = 0x1000
      trace_test(6)
   end,
}

run.L = {
   pre = function()
      MEM[0x1000].L = 0x22222222
      CPU.A[6] = 0x1000
      CPU.D[5].L = 0x30303030
   end,
   t = function()
      CPU.D[4].L = 0x22222222
      TEST.RUN(12)
      TEST.CHECK("MEM[0x1000].L", 0x30303030)
   end,
   f = function()
      CPU.D[4].L = 0x55555555
      TEST.RUN(12)
      TEST.CHECK("CPU.D[4].L", 0x22222222)
   end,
   trace = function()
      CPU.A[6] = 0x1000
      trace_test(12)
   end,
}

return run